source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
     }
 }
 
-//Fix return value allways good fo aligment
 static int heap_alloc_internal_check(int16_t alloc_size)
 {
-    int temp_int, temp_sector = 0, temp_size;
-    temp_size = sizeof(int);
+    int block_count = 0;
+
     if (heap_main == 0) {
         heap_failure(NS_DYN_MEM_HEAP_SECTOR_UNITIALIZED);
-        alloc_size = 0;
     } else if (alloc_size < 1) {
         heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
-        alloc_size = 0;
-    }
-
-
-    if (alloc_size) {
-        //
-        temp_sector = alloc_size;
-        temp_int = (temp_sector % temp_size);
-        if (temp_int) {
-            temp_sector += (temp_size - temp_int);
-        }
-
-
-        temp_sector /= temp_size;
-
-        temp_int = (heap_size / temp_size);
-        temp_int -= 2;
-        if (temp_sector > temp_int) {
-            heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
-            temp_sector = 0;
-        }
+    } else if (alloc_size > (heap_size - 2 * sizeof(int)) ) {
+        heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
+    } else if (alloc_size) {
+        block_count = (alloc_size + sizeof(int) - 1) / sizeof(int);
     }
-    return temp_sector;
+    return block_count;
 }
 
-static int8_t ns_sector_validate(int *sector_start, int direction)
+// Checks that block length indicators equals
+// Block has format: Size of data area in words | data area in words | Size of data area in words
+// If Size is negative it means area is unallocated
+// For direction, use 1 for direction up and -1 for down
+static int8_t ns_block_validate(int *block_start, int direction)
 {
     int8_t ret_val = -1;
-    int size_start, size_end;
-    if (direction) {
-        //Up
-        size_start = *sector_start++;
-        if (size_start < 0) {
-            sector_start += -(size_start);
-        } else {
-            sector_start += size_start;
-        }
-        size_end = *sector_start;
+    int *end = block_start;
+    int size_start = *end;
+    if (direction > 0) {
+        end += (1 + abs(size_start));
     } else {
-        size_end = *sector_start--;
-        if (size_end < 0) {
-            sector_start -= -(size_end);
-        } else {
-            sector_start -= size_end;
-        }
-        size_start = *sector_start;
+        end -= (1 + abs(size_start));
     }
-    if (size_start == size_end) {
+
+    if (size_start && size_start == *end) {
         ret_val = 0;
     }
     return ret_val;
 }
 #endif
 
-void *ns_dyn_mem_alloc(int16_t alloc_size)
+#ifdef STANDARD_MALLOC
+#ifdef USE_IAR
```

Comment:
Hmm. Take this out? I don't really believe it can still be required.

Note that it's actually "ifdef STANDARD_MALLOC", so it's not even protecting the real allocator code! When it was originally put in in 2013, it was unconditional or ifndef. Probably working around some error in an earlier version of the real allocator.